oeps/oep-0045-arch-ops-and-config.rst

```diff
@@ -0,0 +1,169 @@
+
+
+.. _native Dockerfile syntax: https://docs.docker.com/engine/reference/builder/#dockerfile-reference
+
+**Docker Images**
+
+edX will provide Docker images for applications that captures the latest code on the master branch as well as images representing named releases.  edX will not provide these images for named releases prior to the acceptance and implementation of this OEP which is Juniper at time of writing.
+
+
+Configuration
+*************
+ 
+**Django settings**
+
+Having a single artifact that runs with different configurations increases stability by improving development parity with other deployment environments.  edX applications already support configuration overrides via a yaml file for production environments, but development and test environments tend to configure the application using different code paths via a settings/devstack.py or settings/test.py file.
+
+Additionally it is not clear which settings are required to be overridden and which settings have values that may technically work but are inappropriate for production systems.  To alleviate these issue edX django applications will adopt the following settings structure:
+
+.. code-block:: text
+
+  settings
+  ├── __init__.py
+  ├── required.py
+  └── defaults.py
+
+
+*  ``__init__.py`` - Sourcing our config from this file within the settings directory takes advantage of django defaults and means that settings will be picked up automatically without needing to specify ``--settings`` anywhere.  This entry point would import ``required.py``, ``defaults.py``, and the code to override both from a config file.
+*  ``required.py`` - all settings which are required to run and do not have a reasonable production-ready default, e.g. LMS_BASE_URL which will be different per environment.
+*  ``defaults.py`` - all settings other settings which will have production-ready defaults
+
+The settings defined in ``required.py`` and ``defaults.py`` files are mutually exclusive, representing all application specific settings as well as installed library settings whose values either must be provided or whose defaults are not considered production-ready.
+
+``required.py`` variables will all be initialized to ``None`` and the application will not start unless they are set.  This allows operators to fail fast rather than finding out about an unset value when users exercise those breaking codepaths. Application developers are encouraged to keep the list of required settings to a minimum.
+
+This new settings structure obviates the need for any other python files in the settings directory (such as ``devstack.py``, ``test.py``, etc).  The values currently set in those files should be moved to a corresponding ``devstack.yml``, ``test.yml``, etc in the same settings directory.  This gives developers and operators more consistency across environments since the same code paths are being executed with different values.
```

Comment:
Are we talking about using the Docker images for local development? I think we should: or at least, we should be able to test the Docker images locally by pointing to localhost (and not S3, for instance).

If we are to use these Docker images for local development, I'm afraid this is an overly optimistic approach for organising the settings. I don't think we will be able to reach this state where we only have defaults.py/required.py without having to resort to huge prod.yml/dev.yml files.

My point is that `required.py` and `defaults.py`, by themselves, will not be usable in either a development or production environment without very large prod.yml/dev.yml files. The reason for that is that prod.py/dev.py contain not only configuration **values**, but also configuration **logic**. 

Let's take a very simple example: in production we need to set `DEBUG = False`. This is just a fact that is true for all users. Does that mean that _all users_ should now add `DEBUG: false` to their yaml files? I don't think it should: there are quite literally hundreds of settings that need to be set separately for production and development, but for which we can easily predict sane defaults.

The alternatives are:
1. Use a single settings file that differentiates between dev and prod environments. This is the approach used by Sentry. See: https://github.com/getsentry/sentry/blob/685427a5dea1e1ad0adb1816256e28f4958f570f/src/sentry/conf/server.py#L68
2. Use separate settings files: base.py, production.py, development.py. This is advocated by [Two Scoops of Django](https://www.amazon.com/Two-Scoops-Django-Best-Practices/dp/0981467342) and it's Jacob Kaplan-Moss'  ["The One True Way"](https://www.slideshare.net/jacobian/the-best-and-worst-of-django/51).

In the case of edx-platform, I would argue for approach 2.

We would have, similar to what we currently do:

base.py:

```
# Reasonable settings
additional_settings = yaml.load(os.environ["CONFIG_PATH"]) if "CONFIG_PATH" in os.environ else {}
DEBUG = additional_settings.get("DEBUG", False)
ALLOWED_HOSTS = additional_settings.get("ALLOWED_HOSTS", ["localhost", "localhost:8000"])
...
```

development.py:

```
from .base import *

# Development logic
DEBUG = True
ALLOWED_HOSTS = ["*"]
SECRET_KEY = "v3rys3cr3t"
```

production.py:

```
```
from .base import *

# Development logic
DEBUG = False
ALLOWED_HOSTS = [additional_settings["HOST"]]
SECRET_KEY = additional_settings["SECRET_KEY"] # fail fast in case this is not defined
...
```

I understand that, by doing so, we lose the benefit of automatically loading the settings from __init__.py, but in my experience this is not a problem: settings will be automatically loaded as soon as we set the [DJANGO_SETTINGS_MODULE](https://docs.djangoproject.com/en/3.0/topics/settings/#envvar-DJANGO_SETTINGS_MODULE) environment variable (which should be set to "prod" by default in the Docker image)

Furthermore, users will still be able to customize production/development settings by creating their own settings files which will import from production.py/development.py (as we do today). This is pretty much the standard way to proceed in the django world.

Reply:
One of the goals of this OEP is to have an application that executes the same codepaths in development as in production wherever reasonable.  I believe that this is a place that could reasonably be one code path.  Let's take the dev defaults case, because it's a very valid point.  Here is how I would see that working.

required.py
```
# Variables required to run, but without reasonable production defaults
LMS_BASE_URL = None
JWT_AUTH = None
...
```

defaults.py
```
# Variable with reasonable defaults
DEBUG = false
DEFAULT_PAGE_SIZE = 25
STATIC_URL = '/static/'
...
```

__init__.py
```
import os

from .required import *
from .defaults import *

def get_env_setting(setting):
    """ Get the environment setting or return exception """
    try:
        return os.environ[setting]
    except KeyError:
        error_msg = u"Set the %s env variable" % setting
        raise ImproperlyConfigured(error_msg)

CONFIG_FILE = get_env_setting('APP_CFG')

with open(CONFIG_FILE, encoding='utf-8') as f:
    config_from_yaml = yaml.safe_load(f)
    vars().update(config_from_yaml)
```

development.yml (lives in the repo either like this or as starting yaml to copy `development.yml.example`)
```yaml
# required
LMS_BASE_URL: localhost:18000
JWT_AUTH: { ... auth ... }

# default overrides
DEBUG: true
DEFAULT_PAGE_SIZE: 1000
```

production.yml (managed securely elsewhere)
```yaml
# required
LMS_BASE_URL: www.my-prod-lms.com
JWT_AUTH: { ... prod auth ... }
```

Reply:
I am all for giving developers and operators as many of the values defaulted to suit their environments as possible.  It is also likely that for testing we will need/want a `test.yml` to override certain settings.  That should also live in the repo.

IMO this method increases the likelihood that new settings end up in the right place, documented, with proper defaults. And that when people are writing and testing their apps they aren't surprised by how it works in production.